### ISSUE TITLE
Pet follow movement should be head on with pet owner

### ISSUE THREAD

**Issue by Ovahlord** [TrinityCore/TrinityCore#21237]
**Description:**
Right now there is an ancient issue with the follow behaivior of pets. On retail a pet is following you head on. So if you start moving the pet will immediately follow you so it is always at the same level as the owner. On TC however the follow movement has a slight delay. The pet starts following you when you are a few steps ahead. Probably the follow movement generator has a tolerance of several yards which should not apply for pets.

CHANGEME Description of the problem or issue here.

**Current behaviour:**
Pets are not following the pet owner head on:
![wowscrnshot_011518_174211](https://user-images.githubusercontent.com/18347559/34952884-78c5b2e2-fa1b-11e7-9317-129529f08533.jpg)

**Expected behaviour:**
Pets should always be head on head with their pet owner.

**Steps to reproduce the problem:**

1. Summon a pet
2. Start moving
3. Notice the issue

**Branch(es):**
 both

**TC rev. hash/commit:** ff62f22443b0793bbbf6d83d038aa3a19bb840b7

**Comment by Shauren** [comment]
Thats not a delay, on retail your own position is calculated a few steps ahead, you can observe this by running in place into closed gates inside arathi basin

**Comment by Ovahlord** [comment]
Allright. And what may cause the issue that the pet starts following you that late on TC atm? Is it really a tolerance of the follow movement generator?
Cuz in this case things seem to be summed up to:
The follow movement generator has to trigger right when the movement starts with a guessed position (sth like cos(angle)*predictiondistance / sin(angle)...) and also then there most be a immediate stop movement call when the player stops moving. Is this correct?

**Comment by Krudor** [comment]
This is just another issue that chains into other issues in an endless list of issues that no one has time for :P

This logic should also apply to unit formations (Players can also be participants in creature formations, not supported on TrinityCore.)

Also missing proper speed propagation depending on formation leaders movement speed, distance to formation leader etc.

**Comment by chaodhib** [comment]
Yes, there is indeed a [delay](https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Movement/MovementGenerators/TargetedMovementGenerator.cpp#L67). It should be interesting to lower it and see the result.

**Comment by Ovahlord** [comment]
https://youtu.be/WabuFT3N2IE?t=8m17s  this shows pretty well that the pet immediatly follows. While it can also happen that the pet falls back it is mostly because of the pathfinding or if the player turns arround and starts walking.

**Comment by Killyana** [comment]
In retail the position of the pet is calculated a few steps ahead, so it could end up ahead  when the master stop and in this case it will relocate 2 secs later. (not sure if we need such behavior)

The issue is in 3.3.5 is the follow angle has kind of tolerance marge and this will always keep the pet behind the master 290° instead of 270°

I noticed this issue doesn't happen in 4.3.4 and master https://imgur.com/a/bQgLQwt

**Comment by lineagedr** [comment]
Corecraft had retail like pet movement where the pet was always with the owner.

Might be useful:
https://github.com/ccshiro/corecraft/blob/master/server/src/game/Unit.cpp#L3388
https://github.com/ccshiro/corecraft/blob/master/server/src/game/movement/TargetedMovementGenerator.cpp#L745
https://github.com/ccshiro/corecraft/blob/master/server/src/game/PathFinder.cpp#L1064

**Comment by Ovahlord** [comment]
I think the first issue is the way how splines usually are calculated on retail compared to TC. TC stops its spline point generation right at the destination while retail lets the path go a bit ahead of the destination:
![path generation](https://user-images.githubusercontent.com/18347559/69234042-6b333100-0b8e-11ea-96cb-8f70d6f8e4de.png)

**Comment by illfated** [comment]
Thank you for lots of interesting input & feedback. With regards to this issue and the difference between observed movement in 3.3.5a versus 4.3.4 or `master` branch client, is the main difference caused by the source code, or does the game client also play an important part in creating the difference?

**Comment by Ovahlord** [comment]
The client itself should not be a factor in this. The current primary issue is that the follow movement right now is having a certain tolerance that allows the pet to fall behind and not catch up completely to its original follow angle because it only moves to the very next point within the tolerance section

**Comment by Killyana** [comment]
@Ovahlord Do you remember the issue about formations end up broken (wrong follow angle) when the master turns? It's probably caused also by this tolerance value.

**Comment by Killyana** [comment]
As you can see tolerance was added here: https://github.com/TrinityCore/TrinityCore/commit/2a84562dc85516f432bb1e5de9add23c28b26ce4#diff-1d63ea3dcfdaacb4f86c3854ba23bc5dR88-R116
https://github.com/TrinityCore/TrinityCore/commit/2a84562dc85516f432bb1e5de9add23c28b26ce4#diff-417e651d54a4c20c03f76ceaaecf5486R45

**Comment by Ovahlord** [comment]
The formation movement is completely outside of the scope of the issue. Formations have a own movement handling that does not have a movment generator (yet)

**Comment by lineagedr** [comment]
Any news on this?

**Comment by Faq** [comment]
ref https://github.com/TrinityCore/TrinityCore/pull/24840#issuecomment-646286639
